### DISCUSSION TITLE
blog/type-safe-react-query

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#91] [Announcements]
# blog/type-safe-react-query

About the difference between "having types" and "being type-safe"...

https://tkdodo.eu/blog/type-safe-react-query

**Comment by joaooliveiradev** [comment]
Very good article, in my project's i use io-ts to the runtime validation and i also implemented an wrapper of the fetcher that receives an codec and validate inside the fetcher, that way i can guarantee the type-safety in a certain way.

**Comment by developerdanwu** [comment]
Thanks for the article and tips! I’ve been typing useMutation using generics to type the params I need to pass to the mutate fn and possibly response. Wondering, is there a better way you would recommend doing this?

**Reply by TkDodo** [reply]
Same as with useQuery: if the function you pass to useMutation (the mutationFn) has a typed input and output, we don't need to pass generics to useMutation as those will be inferred.

**Comment by Reza-tm** [comment]
Hello tkdodo
I really enjoyed your article.
I'm from iran and i wana help iranian people by translating this article . 
Can i translate this beautiful article and then add Persian translation to this page ?

**Reply by TkDodo** [reply]
Thank you ❤️ . Absolutely, translations are very welcome. Have a look at the contribution guide:

https://github.com/TkDodo/blog/blob/main/CONTRIBUTING.md#translations

**Comment by f0-x** [comment]
Hello there, very helpful article as usual. So, I'm having hard time understanding key differences between tRPC & zodios. Could you please explain in brief when would you pick tRPC or zodios ? Thanks :)

**Reply by TkDodo** [reply]
`tRPC` is perfect if you have the backend under your control, e.g. in the same monorepo, written in TypeScript. `zodios` shines when you communicate with APIs that are not written in TS. I think @ecyrbe, creator of `zodios`, even uses it with `tRPC`: https://twitter.com/ecyrbedev/status/1610622004524097539

**Reply by ecyrbe** [reply]
What i use is not for every project though. 
it's a complex architecture that allows for scallability and partitionning of development and parrallelise work between multiple teams efficiently. 
We could have used GraphQL instead of tRPC in the following schema, but have settled to use tRPC for productivity reasons.
Here is the architecture i'm using. tRPC makes calls to some micro-services using zodios.
Be aware that our micro services are really dumb and don't contain business code. All business logic is on tRPC.
I'll try to make blog about this when i have time.
![image](https://user-images.githubusercontent.com/633115/212472455-fcc23489-c6d4-4cc9-98bd-4650dd9dcdf0.png)

**Comment by robcaldecott** [comment]
This is great, but what about typing the kind of error thrown? For example, I am using `ky` which throws a `HTTPError` so I need to do something like this:

```tsx
const query = useQuery<Todo, HTTPError>({
  queryKey: ['todos', id],
  queryFn: () => ky.get(`/todos/${id}`).json(),
})

// query.error = null | HTPPError
```

Any best practices recommended here instead of defaulting to `unknown` for errors?

**Reply by TkDodo** [reply]
we will default to `Error` in v5. I don't think there's any better way because there could still be other errors thrown, e.g. if you use `select` and you have a runtime error there.

since `HTTPError extends Error`, you can make an `instanceof` check:

```
if (query.error instanceof HTTPError)
```

you need some form of check anyways because you need to eliminate `null`, which this check does as well.

**Comment by timkindberg** [comment]
This is great, but what about this scenario, where I want a wrapper function to accept the proper QueryOptions? How do I use inference here? 

```tsx
type Todo = { id: number; name: string; done: boolean }

const fetchTodo = async (id: number): Promise<Todo> => {
  const response = await axios.get(`/todos/${id}`)
  return response.data
}

//                                 Here I've added `opts`
//                                 Is this how I should type it? I'm forced to use the generics though...
function useTodoQuery(id: number, opts: QueryOptions<Todo>) {
  return useQuery({
    queryKey: ['todos', id],
    queryFn: () => fetchTodo(id),
    ...opts
  })
}
```

**Reply by TkDodo** [reply]
my take on this has _always_ been that you likely do not want to create an abstraction like `useTodoQuery` where you allow to pass in _all options_. It's a very low-level custom hook to build. The standard example is: With this, users can pass in custom cacheTime for every usage, which is likely not what you want. With just `QueryOptions`, they can also pass a different QueryKey:

```
useTodoQuery(1, { queryKey: ['not', 'todos'] })
```

🤷 

so really, a `useTodosQuery` hook doesn't need all options. If you really build something low-level, like a `useQueryWithApiToken` hook or something, then you need to dig into generics. A lot. I have a twitter thread on this: https://twitter.com/TkDodo/status/1491451513264574501

**Reply by timkindberg** [reply]
Ok so... that's fine for like `enabled`, but often we do need to pass at least `onSuccess`... which needs generics right? Or at least to add a type for the argument coming into the onSuccess.

**Reply by TkDodo** [reply]
if you can pass `opts: QueryOptions<Todo>`, you can also pass: `onSuccess: (data: Todo) => void`

**Reply by timkindberg** [reply]
True. Probably the best answer is to not abstract by wrapping the function but rather to abstract by creating query object factories. Like this approach here: https://github.com/TanStack/query/discussions/3888

we have yet to move to this pattern but it would solve this particular pain point because there wouldn’t be the extra layer between using useQuery.

**Comment by ianmacartney** [comment]
I just wrote a post partially inspired by this one about using zod with Convex, which is a backend-as-a-service that shares types between the frontend and backend by using typescript everywhere, all the way to the database schema. Even with end to end typing, though, validation is still important. 
My post is about validating what you pass to the server, so might be a good complement to this post:
https://stack.convex.dev/wrappers-as-middleware-zod-validation

Interested what you think!

**Reply by geraldodev** [reply]
@tkdodo Why axios ?

**Reply by TkDodo** [reply]
I'm mainly using `axios` in the examples because it automatically rejects non-200 status codes, which is in-line with what react-query expects. It's such a common gotcha with the fetch api + react-query. If I were to use plain fetch, I would need to add a couple of distracting lines of boilerplate code. I would like to just use `ky`, which is a bit like axios but based on fetch, however, I fear that it is less well-known and would throw people off.

**Comment by jintaokoong** [comment]
I absolutely like the idea of validating the response data from the backend, but I think there is a difference between network error and validation error. React Query will treat errors thrown within the queryFn to be a network error which a validation error does not represent. A way to go about this is to create some sort of wrapper library that wraps around React Query to preprocess the return value before returning it back to the consumer. It could make the distinction between network and validation errors. Also eliminating the retry problem at the same time.

**Reply by TkDodo** [reply]
> React Query will treat errors thrown within the queryFn to be a network error which a validation error does not represent.

React Query doesn't know about any kinds of errors (network error or non-network error) - it just knows Promises - resolved or rejected. So no, it doesn't treat errors thrown within the queryFn as network errors.

It's true that per default, it treats all errors as equal and will thus trigger `retries`, but you can definitely customize that to e.g. not trigger a retry on a `ZodError`.

**Reply by jintaokoong** [reply]
> React Query doesn't know about any kinds of errors (network error or non-network error) - it just knows Promises - resolved or rejected. So no, it doesn't treat errors thrown within the queryFn as network errors.

I assumed it treats it as network error just because I notice it triggers the retries whenever an error is thrown. My bad for misunderstanding. If you don't mind, would you be so kind to elaborate further on how to customize the retry behavior? Do you mean that we can handle it manually to wrap the result in some sort of structure to prevent the automatic retry or is there an option provided by React Query?

**Reply by ecyrbe** [reply]
it's in the docs, you can provide a custom function to decide if you want to retry the query :  
```ts
{
  retry: boolean | number | (failureCount: number, error: TError) => boolean
}
```

**Reply by jintaokoong** [reply]
@ecyrbe I must have missed this, thanks for the info!

**Comment by ringuralte** [comment]
Thank you for the article. I have a couple of questions with regards to zod. Should we create multiple schemas for the different method calls like GET and POST? In your example you're using the todoSchema to validate GET responses from the api but if I want to validate a POST request for my form, I won't require the `id` field for example since that'll be created by the server. So should I create a different schema to validate the TODO form instead of making the `id` field as optional so they can share the schema?

Also do you bother/recommend changing the casing? Like most api's use snake_casing but camelCase is usually used in the frontend. For example in flutter/dart you can use the json_serializable package to transform snake_case keys to camelCase keys when defining your data model.

**Reply by ecyrbe** [reply]
you can reuse schemas and use `schema.omit({ id: true })`

**Reply by TkDodo** [reply]
yeah I'd usually compose schemas, so if a schema has too many attributes, I'll create another one that `omits` some of the fields. Or, if it's the other way around, you can use `extend`:

```
const DogWithBreed = Dog.extend({
  breed: z.string(),
})
```

**Comment by MarcoB-Ec** [comment]
Hey! Thanks for another insightful article. 

I've been using React Query as a state manager, inspired by one of your articles.

Now, there are cases where a parent component already retrieved the data. And we have applied some checks, like showing something while the data is being fetched. So, in the children component we know the data is defined already, cause the hook will get it from the cache. But it is annoying that TS is telling data could be undefined. It makes sense cause that is the type inferred by useQuery.

So the question: How can we write a custom hook that when I know the data is defined, I can pass a Type Argument to the hook, and the data is type Todo,  not undefined anymore. If I don't pass anything, then it is Todo | undefined. 

Thanks for the help!

**Reply by TkDodo** [reply]
The thing about "knowing" that data will be defined because you've checked it higher up the tree is that it's a bit brittle, because you couple yourself to your current project structure. The child component is, in itself, self-contained. If you move it / refactor it to a different part of your App, data might not be there.

The coupling is implicit: You only know it if you know how the components are structured. On larger teams, I wouldn't trust that by looking at a single component, everyone knows that data will be fetched 5 layers above. If you really want that, I suggest to just use the bang operator (`!`) and shut up typescript - at least that's explicit, and you might have to disable the lint rule that tells you to not use the bang operator because it's unsafe, but what you are doing is essentially unsafe, so at least disabling it is explicit. Or you can write an `invariant` or `assert` function in your custom hook that throws if data is not defined, so you'll instantly know if something goes wrong.

What you can otherwise do is for those cases, not call `useQuery` again, but put data from the parent where it is fetched in react-context. The child would then read from context rather than `useQuery`. Note that this is _not_ state-syncing, but just a form of avoiding prop drilling: The single source of truth is still the Query, and context consumers will update accordingly because the only way to write to the context is via the parent that fetches with `useQuery`.

With that, the coupling becomes `explicit`: If you move your child component, you can see that it has a dependency on `useSomeContext()` rather than `useSomeQuery()` and you know that you can't freely move it around - it needs to be below a `SomeContextProvider`. Of course, concerning types, you might have the same / a similar problem: context needs a default value, so that it also works without an explicit provider. What I do then is usually to, again, create a custom hook with an `invariant`:

```
const SomeContext = React.createContext<SomeValue | undefined>(undefined)
export const useSomeContext = () => {
    const value = React.useContext(SomeContext)

    if (typeof value === 'undefined') {
        throw new Error('SomeContext: No value has been provided')
    }

    return value
}
```

that will correctly infer the value

**Reply by timkindberg** [reply]
Love this reply! Puts into clear writing something I've been vaguely thinking about. Thank you.

**Reply by MarcoB-Ec** [reply]
Amazing! I've learned more about TypeScript, React Query, Design patterns, (I can continue... 😅) in this reply than in any other place since I started my FE journey! Many thanks!

**Comment by chornos13** [comment]
I'am confused about getQueryData example,

if the first example you tell axios signature is a lie if using angle brackets because return-only generic and it is also supported with your argument 
```The more your TypeScript code looks like JavaScript, the better.```

then why in getQueryData example you still using angle bracket over 
```const todo = queryClient.getQueryData(['todos', 1]) as Todo```, 

can you please explain what is the exception using angle bracket when it is return-only generic signature?

**Reply by chornos13** [reply]
*uncompleted 
``` as Todo | undefined ```

**Reply by TkDodo** [reply]
The point is it doesn't matter if you're doing `<Todo>` or `as Todo`. They are both equally bad, though many people think that the generic is fine because it "just specifies a generic", when it's actually the same as a type assertion.

The advantage on `getQueryData` is that we at least add `| undefined` for you. If you use the `as` operator, I would expect mistakes like `as Todo` (as shown in your example) instead of `as Todo | undefined`.

**Comment by metruzanca** [comment]
> It's either unnecessary, or you're lying to yourself.

I love this. When I write generics, I always to try to make them optional. I use them more as a tool / extension to type inference. e.g. think of this function:

```ts
const identity = <T>(data: T): T => data

const i = identity(1)
//    ^? - number
```
`i` is a number, not because we've passed `<number>` to identity but because we've used the generic to tell typescript that it should infer the return type based on the data that was passed in.

**Comment by danbeo95** [comment]
Great!
Thank for your sharing

**Comment by masoud-Tesna** [comment]
Hello, I had to make a custom hook of useMutaion() to execute a code in onSettled. With this, I had a script error on the execution of the mutateAsync function, which said:
Argument type MY_OBJECT is not assignable to parameter type void
That's why I came and used generic type! Can anyone guide me on what to do?
I want a series of toasts to be displayed after useMutation(), which is always fixed, and also the errors or messages that come back from the server, I have to change them, which are fixed in all useMutation()s! That's why I didn't want to repeat these codes every time I use useMutation()!

**Reply by TkDodo** [reply]
please show a typescript playground with your problem

**Comment by debater-coder** [comment]
I used to do zod validation and transformation in the `queryFn`, but since what I was transforming to wouldn't serialise with `JSON.stringify()` and I was using persistent cache, it would cause some issues. I'm considering solving this problem by moving zod validation into `select`. Do you think this would solve that problem?

**Reply by TkDodo** [reply]
I would probably just switch to superjson for the persistent cache then

**Comment by Jamalahmad123** [comment]
Hi @TKDodo, I was stuck on a problem but managed to resolve it. I'd like your opinion on my solution.

The situation is this: I have one API that serves different lists based on an ID. I wanted to create a reusable useQuery hook that can return a generic type. When using the hook, I would pass the specific type I need. Here’s my implementation:

```
export const fetchActivityListById = <T>({
  queryKey: [{ activityId, queryParams }],
}: QueryFunctionContext<
  ReturnType<(typeof activitiesKeys)["list"]>
>): Promise<T> => {
  let endPoint = "events/list/";

  if (queryParams) {
    endPoint += `?activityId=${activityId}&page=${queryParams.pageIndex}&limit=${queryParams.pageSize}&search=${queryParams.search}`;
  }

  return api.get(endPoint);
};

export const useFetchActivityListById = <T>(
  activityId: string,
  queryParams?: QueryParams
): UseQueryResult<PaginatedResponse<T[]>> => {
  return useQuery({
    queryKey: activitiesKeys.list(activityId, queryParams),
    queryFn: fetchActivityListById,
  });
};

const query = useFetchActivityListById<IKHConference>("1")

**Reply by TkDodo** [reply]
The downside is you have to build a manual connection between “id 1” and “the type IKHConference”. It would be great if this could be inferred, e.g. with a Map on type level:

```
type Mapping = {
  “1”: IKHConference
}
```

here’s a quick playground with that concept: https://www.typescriptlang.org/play/?#code/C4TwDgpgBAsghmMBLAdgcygXigbwLABQUUARAIwkBcUKArgLYBGEATgDSHEkBMVUAzsBao0hAL6FCoSFADSEEPyxQA1goD2AM1gJk6KeGgA1OABtaEJdniIRAbTUgtO2+gC6kgptooAxsCR1FChNFAAeeRAoCAAPYAgUABMlSP4APgAKVDBaYGpIgEpqGz00O0i3XE4oFghgWhZgulNTKDglOBQQcU9fIMEoGOVQjPISAsIAekniYgA9AH5e-uAoKOwRnnGpmdnFz02AZm2CIA

**Reply by Jamalahmad123** [reply]
The id will be dynamic and will come from RRD hook useParams. It is not static. here is it
```
const KHExchangeConferences = () => {
  const { activityId } = useParams() as ActivityURLParams;
  const { searchParams, setSearchParams, queryOptions } = usePaginationQuery();

  const { data, isError, error, isPending, isFetching, isSuccess } =
    useFetchActivityListById<IKHConference>(activityId, queryOptions);

  if (isPending) {
    return <div>Loading</div>;
  }

  if (isError) {
    return <ErrorMessage>{error.message}</ErrorMessage>;
  }

  return (
    isSuccess && (
      <ServerDataTable
        columns={columns}
        data={data.data}
        searchParams={searchParams}
        setSearchParams={setSearchParams}
        rowCount={data.pagination.rowCount}
        isFetching={isFetching}
      />
    )
  );
};
```
If you understand it correctly please give me a complete example

**Reply by TkDodo** [reply]
Then I don’t understand why you need a “dynamic” function? I mean, is there a different `activityId` that would _not_ return something of type `IKHConference` ? Or is this like: ids 1-25 are this type, and ids 26-50 are a different type ? That would be a weird API 🤷‍♂️

**Reply by Jamalahmad123** [reply]
Yes of course API return different type like here
```
//************* Activity **************/
export type EventActivity = {
  id: number;
  title: IActivity;
  activityType: IActivityType;
};

type ActivityEntity = Entity<{
  title: string;
  date: string;
  location: string;
  activity: EventActivity;
  status: IStatuses;
  subactivity: {
    id: number;
    title: string;
  } | null;
}>;

//************ Meeting ***********/

export type IMeeting = ActivityEntity & {
  mode: InteractionType;
  hostCountry?: string;
  requestedCountry?: string;
};


//************ Events ***********/

export type IEvent = ActivityEntity & {
  mode: InteractionType;
  hostCountry: string;
};

//************ Conferences ***********/

export type IKHConference = ActivityEntity & {
  mode: InteractionType;
};

//************ Training and Workshop ***********/

export type ITrainingAndWorkshop = ActivityEntity & {
  mode: InteractionType;
};

//************ Research and Foresight ***********/

export type IResearchAndForesight = ActivityEntity & {
  mode: InteractionType;
};
```
some of the different data will be in the details of each activity

**Comment by cmacdonnacha** [comment]
Hey, really appreciate this article. In 2026, which of the 2 options below would you prefer? Let's say Zod is out of the picture for this particular case.

Type the function:

```ts
const fetchTodo = async (id: number): Promise<Todo> => {
  const response = await axios.get(`/todos/${id}`)
  return response.data
}
```

Type the response:

```ts
const fetchTodo = async (id: number) => {
  const response = await axios.get<Todo>(`/todos/${id}`)
  return response.data
}
```

**Reply by cmacdonnacha** [reply]
I feel like typing `axios.get` is best since it gives you bets of both worlds:

- tells the caller what to expect
- gives type safety within the function

**Reply by TkDodo** [reply]
I probably wouldn’t use `axios` in 2026. More modern solutions like [ky](https://github.com/sindresorhus/ky) return `unknown` by default instead of `any`, in which case solution 1 doesn’t work if you have additional data transformations inside your `queryFn`. With `axios`, the range of `any` is larger in solution 1, so those transformations would not be properly typed:

```
const fetchTodo = async (id: number): Promise<Todo> => {
  const response = await axios.get(`/todos/${id}`)
  // response.data is any so we just pass any here
  return transformation(response.data)
}
```

With `ky`, that data would be `unknown` and we would be forced to do the assertion “inside”:

```
const fetchTodo4 = async (id: number) => {
  const response = await ky.get<Todo>(`/todos/${id}`)
  const data = await response.json()
  return transformation(data)
}
```

so with that in mind, I think I’d do the second version.

**Reply by cmacdonnacha** [reply]
Really appreciate the reply! I'd love to use something like Orval to auto generate the clients and queries for me, but I did notice that the generated output always has the function response typed: https://orval.dev/playground

In this case I'd wonder if there would be an issue with using `ky` with Orval, perhaps that's where we'd have to use a custom http client config.
